File: app/src/main/java/com/huck/biblequiz/util/QuizGenerator.kt
package com.huck.biblequiz.util

import com.huck.biblequiz.model.Blank
import com.huck.biblequiz.model.QuizQuestion
import com.huck.biblequiz.model.Verse

object QuizGenerator {

    private val stopWords = setOf(
        "the", "and", "that", "this", "with", "from", "your", "have", "has",
        "his", "her", "they", "them", "their", "been", "were", "was", "are",
        "but", "not", "you", "all", "can", "had", "who", "will", "for",
        "shall", "unto", "upon", "said", "also", "into", "than", "then",
        "when", "which", "whom", "what", "there", "would", "could", "should"
    )

    fun generate(verses: List<Verse>): List<QuizQuestion> {
        return verses.mapNotNull { verse ->
            generateQuestion(verse)
        }
    }

    private fun generateQuestion(verse: Verse): QuizQuestion? {
        val words = verse.text.split("\\s+".toRegex()).filter { it.isNotBlank() }
        if (words.size < 3) return null

        val blankCount = when {
            words.size < 8 -> 1
            words.size <= 15 -> (1..2).random()
            else -> (2..3).random()
        }

        val candidates = words.indices.filter { i ->
            val word = stripPunctuation(words[i])
            word.length >= 4 && word.lowercase() !in stopWords
        }

        if (candidates.isEmpty()) return null

        val scored = candidates.sortedByDescending { i ->
            val word = stripPunctuation(words[i])
            var score = word.length
            if (i > 0 && word[0].isUpperCase()) score += 3
            score
        }

        val selected = scored.take(blankCount * 2)
            .shuffled()
            .take(blankCount)
            .sorted()

        val blanks = selected.map { i ->
            Blank(index = i, answer = stripPunctuation(words[i]))
        }

        val displayWords = words.toMutableList()
        for (blank in blanks) {
            val original = displayWords[blank.index]
            val trailingPunct = original.takeLastWhile { !it.isLetterOrDigit() }
            displayWords[blank.index] = "______$trailingPunct"
        }

        return QuizQuestion(
            verse = verse,
            displayText = displayWords.joinToString(" "),
            blanks = blanks
        )
    }

    fun checkAnswer(userAnswer: String, correctAnswer: String): Boolean {
        return stripPunctuation(userAnswer.trim()).equals(
            stripPunctuation(correctAnswer.trim()),
            ignoreCase = true
        )
    }

    private fun stripPunctuation(word: String): String {
        return word.trimEnd { !it.isLetterOrDigit() }.trimStart { !it.isLetterOrDigit() }
    }
}

File: ios/BeSpecific/Utilities/QuizGenerator.swift
import Foundation

enum QuizGenerator {
    private static let stopWords: Set<String> = [
        "the", "and", "that", "this", "with", "from", "your", "have", "has",
        "his", "her", "they", "them", "their", "been", "were", "was", "are",
        "but", "not", "you", "all", "can", "had", "who", "will", "for",
        "shall", "unto", "upon", "said", "also", "into", "than", "then",
        "when", "which", "whom", "what", "there", "would", "could", "should"
    ]

    static func generate(verses: [Verse]) -> [QuizQuestion] {
        verses.compactMap { generateQuestion($0) }
    }

    private static func generateQuestion(_ verse: Verse) -> QuizQuestion? {
        let words = verse.text.components(separatedBy: .whitespaces).filter { !$0.isEmpty }
        guard words.count >= 3 else { return nil }

        let blankCount: Int
        switch words.count {
        case ..<8: blankCount = 1
        case 8...15: blankCount = Int.random(in: 1...2)
        default: blankCount = Int.random(in: 2...3)
        }

        let candidates = words.indices.filter { i in
            let word = stripPunctuation(words[i])
            return word.count >= 4 && !stopWords.contains(word.lowercased())
        }

        guard !candidates.isEmpty else { return nil }

        let scored = candidates.sorted { a, b in
            let wordA = stripPunctuation(words[a])
            let wordB = stripPunctuation(words[b])
            var scoreA = wordA.count
            var scoreB = wordB.count
            if a > 0, let first = wordA.first, first.isUppercase { scoreA += 3 }
            if b > 0, let first = wordB.first, first.isUppercase { scoreB += 3 }
            return scoreA > scoreB
        }

        let selected = Array(scored.prefix(blankCount * 2))
            .shuffled()
            .prefix(blankCount)
            .sorted()

        let blanks = selected.map { i in
            Blank(index: i, answer: stripPunctuation(words[i]))
        }

        var displayWords = words
        for blank in blanks {
            let original = displayWords[blank.index]
            let trailingPunct = String(original.reversed().prefix(while: { !$0.isLetter && !$0.isNumber }).reversed())
            displayWords[blank.index] = "______\(trailingPunct)"
        }

        return QuizQuestion(
            verse: verse,
            displayText: displayWords.joined(separator: " "),
            blanks: blanks
        )
    }

    static func checkAnswer(_ userAnswer: String, _ correctAnswer: String) -> Bool {
        removePunctuation(userAnswer.trimmingCharacters(in: .whitespaces))
            .caseInsensitiveCompare(
                removePunctuation(correctAnswer.trimmingCharacters(in: .whitespaces))
            ) == .orderedSame
    }

    private static func stripPunctuation(_ word: String) -> String {
        var result = word
        while let last = result.last, !last.isLetter, !last.isNumber {
            result.removeLast()
        }
        while let first = result.first, !first.isLetter, !first.isNumber {
            result.removeFirst()
        }
        return result
    }

    private static func removePunctuation(_ text: String) -> String {
        let cleaned = text.unicodeScalars.filter {
            CharacterSet.letters.contains($0) || CharacterSet.decimalDigits.contains($0) || CharacterSet.whitespaces.contains($0)
        }
        let result = String(String.UnicodeScalarView(cleaned))
        return result.components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespaces)
    }
}
